### PR TITLE
Adding `$merge`

### PR DESCRIPTION
**Pull request by TheNavigat**
Closes https://github.com/taskcluster/json-e/issues/59

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -105,6 +105,26 @@ constructs.$map = (template, context) => {
 
 };
 
+constructs.$merge = (template, context) => {
+  let value = render(template['$merge'], context);
+
+  if (!isArray(value) && !isArray(template['$merge'])) {
```
Comment (jonasfj) [inline src/index.js]: Explain why you have `!isArray(template['$merge'])`

File: src/index.js [hunk 2]
```diff
@@ -105,6 +105,26 @@ constructs.$map = (template, context) => {
...
+
+  if (!isArray(value)) {
+    throw jsonTemplateError('$merge requires array as value\n', template);
+  }
+
+  let obj = {};
+
+  value.forEach(function(x) {
+    Object.assign(obj, x);
+  });
```
Comment (jonasfj) [inline src/index.js]: Can't you do: `return Object.assgin({}, ...value);` or `return Object.assgin({}, ...value.reverse());` (you probably need to check semantics of `Object.assign`

File: src/index.js (hunk 1 above)
Comment (jonasfj) [inline src/index.js]: IMO we only care what it evaluates to... not how it got to be that value...

Comment (TheNavigat) [inline src/index.js]: You're correct, I deleted it that check.

File: src/index.js (hunk 2 above)
Comment (TheNavigat) [inline src/index.js]: I'm not sure how to do that. I checked `Object.assign`'s documentation and couldn't find any information on how to assign an array of objects in a single line.

When I tried, this happened.

```javascript
> x = [{a: 1}, {a: 2}, {a: 3}]
[ { a: 1 }, { a: 2 }, { a: 3 } ]
> Object.assign({}, x);
{ '0': { a: 1 }, '1': { a: 2 }, '2': { a: 3 } }
```

**Comment by jonasfj** [comment]
`Object.assign({}, a, b, ...c)` is valid js... So is where c is a list of
objects...

Den 21. maj 2017 7.05 PM skrev "Amjad Mashaal" <notifications@github.com>:

**Comment by jonasfj** [comment]
As you can see here sources can be more than one:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

Den 21. maj 2017 8.49 PM skrev "Jonas Finnemann Jensen" <jopsen@gmail.com>:

**Comment by TheNavigat** [comment]
@jonasfj If c is a list that doesn't work, nope.

`Object.assign({}, a, b, c);` works but I'm not sure how can I execute the command with `a, b, c` where those arguments are actually a list. Can you show me how to do that, please?

**Comment by jonasfj** [comment]
`Object.assign({}, ...[a, b, c]);`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator

**Comment by TheNavigat** [comment]
I never knew that existed. I assumed the `...` you typed above meant the `...` used in English, not as an operator. I apologize :)

Code is updated. Please let me know if there's anything I should fix/improve.

File: README.md [hunk 3]
```diff
@@ -167,9 +167,17 @@ template:
 result:   [3, 5, 7]
 ```
 
-The array is the value of the `$map` property, and the expression to evaluate
-is given by `each(var)` where `var` is the name of the variable containing each
-element.
```
Comment (djmitche) [inline README.md]: This paragraph of documentation for `$map` appears to have disappeared..

File: specification.yml [hunk 4]
```diff
@@ -55,6 +55,11 @@ context:  {}
 template: {$json: [1,2,true,{},[]]}
 result:   '[1,2,true,{},[]]'
 ---
+title:    $merge
+context:  {}
+template: {$merge: [{a: 1, b: 1}, {b: 2, c: 3}, {d: 4}]}
+result:   {a: 1, b: 2, c: 3, d: 4}
+---
```
Comment (djmitche) [inline specification.yml]: We should probably test some edge cases, too.  What does `{$merge: []}` return?  How about `{$merge: [{}]}`?
